### PR TITLE
Bug 2094438: Make AWS URL parsing more lenient for GetNodeEgressIPConfiguration

### PR DESCRIPTION
**Pull request by andreaskaris**
Signed-off-by: Andreas Karis <ak.karis@gmail.com>

For context, see: https://bugzilla.redhat.com/show_bug.cgi?id=2092003

### CODE REVIEW

**Review by flavio-fernandes — COMMENTED** [review]
/lgtm

**Comment by andreaskaris** [comment]
~~~
[akaris@linux kubernetes-tools (master)]$ export KUBECONFIG=~/cases/kubeconfig
[akaris@linux kubernetes-tools (master)]$ oc get nodes
NAME                                         STATUS   ROLES    AGE   VERSION
ip-10-0-129-211.us-east-2.compute.internal   Ready    master   58m   v1.23.3+ad897c4
ip-10-0-135-193.us-east-2.compute.internal   Ready    worker   51m   v1.23.3+ad897c4
ip-10-0-142-80.us-east-2.compute.internal    Ready    master   58m   v1.23.3+ad897c4
ip-10-0-152-244.us-east-2.compute.internal   Ready    worker   52m   v1.23.3+ad897c4
ip-10-0-198-143.us-east-2.compute.internal   Ready    master   58m   v1.23.3+ad897c4
ip-10-0-208-122.us-east-2.compute.internal   Ready    worker   52m   v1.23.3+ad897c4
[akaris@linux kubernetes-tools (master)]$ oc get co
NAME                                       VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
authentication                             4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      44m     
baremetal                                  4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
cloud-controller-manager                   4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      58m     
cloud-credential                           4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      57m     
cluster-autoscaler                         4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
config-operator                            4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
console                                    4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      45m     
csi-snapshot-controller                    4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
dns                                        4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
etcd                                       4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
image-registry                             4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      50m     
ingress                                    4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      50m     
insights                                   4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      49m     
kube-apiserver                             4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      47m     
kube-controller-manager                    4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      53m     
kube-scheduler                             4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      53m     
kube-storage-version-migrator              4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
machine-api                                4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      52m     
machine-approver                           4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
machine-config                             4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
marketplace                                4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
monitoring                                 4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      48m     
network                                    4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      57m     
node-tuning                                4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      55m     
openshift-apiserver                        4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      47m     
openshift-controller-manager               4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      53m     
openshift-samples                          4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      50m     
operator-lifecycle-manager                 4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
operator-lifecycle-manager-catalog         4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
operator-lifecycle-manager-packageserver   4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      51m     
service-ca                                 4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
storage                                    4.11.0-0.ci.test-2022-05-31-163144-ci-ln-fxbsdqk-latest   True        False         False      56m     
[akaris@linux kubernetes-tools (master)]$ oc get nodes -o yaml | grep providerID
    providerID: aws://us-east-2b/i-01b190e820dcf1b49
    providerID: aws://us-east-2b/i-0546dcb70e35d1f61
    providerID: aws://us-east-2b/i-00c264a8bf312835f
    providerID: aws://us-east-2b/i-05d5c5e78264ae061
    providerID: aws://us-east-2c/i-020cd7b51f6b2845f
    providerID: aws://us-east-2c/i-00170682b7064c563
[akaris@linux kubernetes-tools (master)]$ oc get pods -A -o wide | grep cloud | grep network
oc logs openshift-cloud-network-config-controller          cloud-network-config-controller-5d64478fc7-bwtlp                            1/1     Running     0             58m   10.128.0.18    ip-10-0-129-211.us-east-2.compute.internal   <none>           <none>
[akaris@linux kubernetes-tools (master)]$ oc logs -n  openshift-cloud-network-config-controller          cloud-network-config-controller-5d64478fc7-bwtlp | tail
I0531 16:46:10.382103       1 controller.go:160] Dropping key 'ip-10-0-129-211.us-east-2.compute.internal' from the node workqueue
I0531 16:49:26.855225       1 controller.go:182] Assigning key: ip-10-0-152-244.us-east-2.compute.internal to node workqueue
I0531 16:49:27.091383       1 node_controller.go:106] Setting annotation: 'cloud.network.openshift.io/egress-ipconfig: [{"interface":"eni-05b80851cee0da570","ifaddr":{"ipv4":"10.0.128.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]' on node: ip-10-0-152-244.us-east-2.compute.internal
I0531 16:49:27.159875       1 controller.go:160] Dropping key 'ip-10-0-152-244.us-east-2.compute.internal' from the node workqueue
I0531 16:49:31.403067       1 controller.go:182] Assigning key: ip-10-0-208-122.us-east-2.compute.internal to node workqueue
I0531 16:49:31.570071       1 node_controller.go:106] Setting annotation: 'cloud.network.openshift.io/egress-ipconfig: [{"interface":"eni-0b3cd757792cba9ef","ifaddr":{"ipv4":"10.0.192.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]' on node: ip-10-0-208-122.us-east-2.compute.internal
I0531 16:49:31.627072       1 controller.go:160] Dropping key 'ip-10-0-208-122.us-east-2.compute.internal' from the node workqueue
I0531 16:49:45.616977       1 controller.go:182] Assigning key: ip-10-0-135-193.us-east-2.compute.internal to node workqueue
I0531 16:49:45.803867       1 node_controller.go:106] Setting annotation: 'cloud.network.openshift.io/egress-ipconfig: [{"interface":"eni-0ca6887a68f12f00c","ifaddr":{"ipv4":"10.0.128.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]' on node: ip-10-0-135-193.us-east-2.compute.internal
I0531 16:49:45.836646       1 controller.go:160] Dropping key 'ip-10-0-135-193.us-east-2.compute.internal' from the node workqueue
[akaris@linux kubernetes-tools (master)]$ oc get nodes -o yaml | grep egress-ipconfig
      cloud.network.openshift.io/egress-ipconfig: '[{"interface":"eni-0552157a3b2bc8b73","ifaddr":{"ipv4":"10.0.128.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]'
      cloud.network.openshift.io/egress-ipconfig: '[{"interface":"eni-0ca6887a68f12f00c","ifaddr":{"ipv4":"10.0.128.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]'
      cloud.network.openshift.io/egress-ipconfig: '[{"interface":"eni-0b402d963b87cb413","ifaddr":{"ipv4":"10.0.128.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]'
      cloud.network.openshift.io/egress-ipconfig: '[{"interface":"eni-05b80851cee0da570","ifaddr":{"ipv4":"10.0.128.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]'
      cloud.network.openshift.io/egress-ipconfig: '[{"interface":"eni-0bc9d1329bd33a2cd","ifaddr":{"ipv4":"10.0.192.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]'
      cloud.network.openshift.io/egress-ipconfig: '[{"interface":"eni-0b3cd757792cba9ef","ifaddr":{"ipv4":"10.0.192.0/18"},"capacity":{"ipv4":14,"ipv6":15}}]'
[akaris@linux kubernetes-tools (master)]$ 
~~~

**Comment by flavio-fernandes** [comment]
tested regexp to make sure it does what is needed and it seems fine:

```
// You can edit this code!
// Click here and start typing.
package main

import (
	"fmt"
	"os"
	"regexp"
)

func main() {

	// Accept the following provider IDs, examples:
	// "url1": "aws:///us-west-2a/i-008447f243eead273",
	// "url2": "aws://us-west-2a/i-008447f243eead273",
	// "url3": "aws:////i-008447f243eead273",
	// "url4": "aws:///i-008447f243eead273".

	vals := []string{
		"aws:///us-west-2a/i-008447f243eead273",
		"aws://us-west-2a/i-008447f243eead273",
		"aws:////i-008447f243eead273",
		"aws:///i-008447f243eead273",
		// additionals
		"aws:///a",
		"aws:////a",

		// BAD:  "aws://a",
	}

	for _, val := range vals {

		re := regexp.MustCompile(`^aws:[/]{2,3}[^/]*/[^/]+$`)
		if !re.Match([]byte(val)) {
			fmt.Println("BROKEN: ", val)
			os.Exit(1)
		}
	}

	fmt.Println("passed")

}
```

**Comment by kyrtapz** [comment]
/lgtm
cloud-provider-aws implementation of the same thing: https://github.com/kubernetes/cloud-provider-aws/blob/5f394ba297bf280ceb3edfc38922630b4bd83f46/pkg/providers/v2/instances.go#L254

**Comment by andreaskaris** [comment]
@flavio-fernandes @kyrtapz I think that's better now, I also added a test case

File: pkg/cloudprovider/aws.go [hunk 1]
```diff
@@ -348,3 +345,26 @@ func (a *AWS) getInstance(node *corev1.Node) (*ec2.Instance, error) {
...
+//      providerID: aws://us-west-2a/i-008447f243eead273
+//      providerID: aws:///i-008447f243eead273
+// see https://github.com/kubernetes/cloud-provider-aws/blob/5f394ba297bf280ceb3edfc38922630b4bd83f46/pkg/providers/v2/instances.go#L254
+func getInstanceIdFromProviderId(providerId string) (string, error) {
+	// after trimming 'aws://', the remainder will be in one of the following formats:
+	// * <availability-zone>/<instance-id>
+	// * /<availability-zone>/<instance-id>
+	// * <instance-id>
+	splitProviderId := strings.Split(strings.TrimPrefix(providerId, "aws://"), "/")
+	if len(splitProviderId) < 1 || len(splitProviderId) > 3 {
```
Comment (kyrtapz) [inline pkg/cloudprovider/aws.go:365]: `splitProviderId` will always have at least one element, https://pkg.go.dev/strings#Split:
`If s does not contain sep and sep is not empty, Split returns a slice of length 1 whose only element is s. `

Comment (andreaskaris) [reply pkg/cloudprovider/aws.go]: Addressed, thanks

Comment (andreaskaris) [reply pkg/cloudprovider/aws.go]: and no, it was just a mistake. Should I check for an empty providerID, though?

Comment (andreaskaris) [reply pkg/cloudprovider/aws.go]: ok I added a check for that as well, PTAL :-)

**Review by flavio-fernandes — COMMENTED** [review]
/lgtm

**Review by trozet — APPROVED** [review]
/approve

**Pull request merged**